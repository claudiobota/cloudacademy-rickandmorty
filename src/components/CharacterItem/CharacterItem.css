.character {
    flex-grow: 1;
    opacity: 0.9;
    transition: opacity 0.2s ease;
    display: flex;
    flex-direction: column;
}

.character .p-card-body {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
}

.character .p-card-body .p-card-content {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
}

.character:hover {
    cursor: default;
    opacity: 1;
}

.character img {
    width: 100%;
    max-width: 200px;
    height: auto;
    margin: 0;
    padding: 0;
    border-radius: 100%;
}

.character .information-item {
    border-bottom: 1px solid rgba(0, 0, 0, 0.05);
}
