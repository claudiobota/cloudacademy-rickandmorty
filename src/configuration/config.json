{
	"api": {
		"url": "https://rickandmortyapi.com/api",
		"paths": {
			"characters": "/character",
			"locations": "/location",
			"episodes": "/episode"
		},
		"page_size": 20
	}
}
