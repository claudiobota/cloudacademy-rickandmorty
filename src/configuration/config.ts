import configData from './config.json';
import configDevelopmentData from './config.development.json';
import configTestData from './config.test.json';
import configStagingData from './config.staging.json';
import configProductionData from './config.production.json';
import {IConfig} from '../types/interfaces';
import {Environment} from '../types/enums';

let config: IConfig = {
  ...configData
};

const nodeEnv: string = process.env.NODE_ENV || 'development';

switch (nodeEnv) {
  case Environment.DEVELOPMENT:
    config = {
      ...config,
      ...configDevelopmentData
    };
    break;

  case Environment.TEST:
    config = {
      ...config,
      ...configTestData
    };
    break;

  case Environment.STAGING:
    config = {
      ...config,
      ...configStagingData
    };
    break;

  case Environment.PRODUCTION:
    config = {
      ...config,
      ...configProductionData
    };
    break;

  default:
}

export default config;
