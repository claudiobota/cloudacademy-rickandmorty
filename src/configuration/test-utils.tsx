import React from 'react';
import {render as rtlRender} from '@testing-library/react';
import {Provider} from 'react-redux';
import {configureAppStore} from '../redux/store';

function render(
  ui: React.ReactElement,
  {
    initialState,
    store = configureAppStore(initialState),
    ...renderOptions
  }: any = {}
) {
  function Wrapper({ children }: { children: React.ReactElement }) {
    return <Provider store={store}>{children}</Provider>;
  }

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { render }
