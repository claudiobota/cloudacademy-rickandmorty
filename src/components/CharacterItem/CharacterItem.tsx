import React, {useRef} from 'react';
import {Card} from 'primereact/card';
import {ICharacter} from '../../types/interfaces';
import './CharacterItem.css';
import {Button} from 'primereact/button';
import CharacterEpisodesDialog from '../CharacterEpisodesDialog/CharacterEpisodesDialog';

export default function CharacterItem(props: { character: ICharacter }) {
  const { character } = props;

  const informationItems = [
    { label: 'Status', key: 'status' },
    { label: 'Species', key: 'species' },
    { label: 'Gender', key: 'gender' },
    { label: 'Type', key: 'type' }
  ]

  const episodesDialogRef = useRef<any>();

  const onViewEpisodesClick = () => {
    episodesDialogRef.current.show();
  }

  return (
    <Card title={character.name} className="character">
      <section className="profile-picture p-text-center">
        <img src={character.image} alt={character.name} />
      </section>

      <section className="information p-mt-5">
        {informationItems.map(informationItem => {
          return (
            <div key={informationItem.key} className="information-item p-pt-1 p-pb-1 p-mb-1">
              <div className="p-grid">
                <div className="p-col-4">
                  <strong>{informationItem.label}</strong>
                </div>
                <div className="p-col-8">
                  <span>{character[informationItem.key as keyof ICharacter]}</span>
                </div>
              </div>
            </div>
          );
        })}

        <div className="information-item p-pt-1 p-pb-1 p-mb-1">
          <div className="p-grid">
            <div className="p-col-4">
              <strong>Origin</strong>
            </div>
            <div className="p-col-8">
              <a href={character.origin.url} target="_blank">{character.origin.name}</a>
            </div>
          </div>
        </div>

        <div className="information-item p-pt-1 p-pb-1 p-mb-1">
          <div className="p-grid">
            <div className="p-col-4">
              <strong>Location</strong>
            </div>
            <div className="p-col-8">
              <a href={character.location.url} target="_blank">{character.location.name}</a>
            </div>
          </div>
        </div>
      </section>

      <div className="p-mt-auto">
        <section className="button-bar p-mt-3">
          <div className="p-grid">
            <div className="p-col-4 p-text-center">
              <Button icon="pi pi-globe" className="p-button-rounded p-button-outlined" tooltip="View location's residents" tooltipOptions={{position: 'top'}} />
            </div>
            <div className="p-col-4 p-text-center">
              <Button icon="pi pi-home" className="p-button-rounded p-button-outlined" tooltip="View origin's residents" tooltipOptions={{position: 'top'}} />
            </div>
            <div className="p-col-4 p-text-center">
              <Button icon="pi pi-external-link" className="p-button-rounded p-button-outlined" tooltip="External Link" tooltipOptions={{position: 'top'}} onClick={() => window.open(`https://rickandmorty.fandom.com/wiki/${character.name.replace(' ', '_')}`, '_blank')} />
            </div>
          </div>
        </section>

        <footer className="p-mt-3">
          <div className="p-grid">
            <div className="p-col-12">
              <Button label="View Episodes" className="p-button-raised p-button-lg p-text-uppercase p-text-bold" style={{ width: '100%' }} onClick={onViewEpisodesClick} />
            </div>
          </div>
        </footer>
      </div>

      <CharacterEpisodesDialog ref={episodesDialogRef} character={character} />
    </Card>
  )
}
