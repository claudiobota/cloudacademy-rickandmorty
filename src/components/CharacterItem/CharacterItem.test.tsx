import React from 'react';
import {fireEvent, render, screen} from '../../configuration/test-utils';
import CharacterItem from './CharacterItem';
import mockData from '../../configuration/mock-data.json';

describe('CharacterItem', () => {
  test('component renders and matches snapshot', () => {
    const {asFragment} = render(<CharacterItem character={mockData.api.characters.results[0]} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('CharacterEpisodesDialog is correctly shown when clicking View Episodes button', () => {
    const {container} = render(<CharacterItem character={mockData.api.characters.results[0]} />);
    expect(container.getElementsByClassName('p-dialog-visible').length).toBeFalsy();
    fireEvent.click(screen.getByText('View Episodes'));
    expect(container.getElementsByClassName('p-dialog-visible').length).toBeTruthy();
  });
});
