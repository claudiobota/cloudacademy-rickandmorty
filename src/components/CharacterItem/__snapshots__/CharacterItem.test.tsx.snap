// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterItem component renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-card p-component character"
  >
    <div
      class="p-card-body"
    >
      <div
        class="p-card-title"
      >
        Rick Sanchez
      </div>
      <div
        class="p-card-content"
      >
        <section
          class="profile-picture p-text-center"
        >
          <img
            alt="Rick Sanchez"
            src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
          />
        </section>
        <section
          class="information p-mt-5"
        >
          <div
            class="information-item p-pt-1 p-pb-1 p-mb-1"
          >
            <div
              class="p-grid"
            >
              <div
                class="p-col-4"
              >
                <strong>
                  Status
                </strong>
              </div>
              <div
                class="p-col-8"
              >
                <span>
                  Alive
                </span>
              </div>
            </div>
          </div>
          <div
            class="information-item p-pt-1 p-pb-1 p-mb-1"
          >
            <div
              class="p-grid"
            >
              <div
                class="p-col-4"
              >
                <strong>
                  Species
                </strong>
              </div>
              <div
                class="p-col-8"
              >
                <span>
                  Human
                </span>
              </div>
            </div>
          </div>
          <div
            class="information-item p-pt-1 p-pb-1 p-mb-1"
          >
            <div
              class="p-grid"
            >
              <div
                class="p-col-4"
              >
                <strong>
                  Gender
                </strong>
              </div>
              <div
                class="p-col-8"
              >
                <span>
                  Male
                </span>
              </div>
            </div>
          </div>
          <div
            class="information-item p-pt-1 p-pb-1 p-mb-1"
          >
            <div
              class="p-grid"
            >
              <div
                class="p-col-4"
              >
                <strong>
                  Type
                </strong>
              </div>
              <div
                class="p-col-8"
              >
                <span />
              </div>
            </div>
          </div>
          <div
            class="information-item p-pt-1 p-pb-1 p-mb-1"
          >
            <div
              class="p-grid"
            >
              <div
                class="p-col-4"
              >
                <strong>
                  Origin
                </strong>
              </div>
              <div
                class="p-col-8"
              >
                <a
                  href="https://rickandmortyapi.com/api/location/1"
                  target="_blank"
                >
                  Earth (C-137)
                </a>
              </div>
            </div>
          </div>
          <div
            class="information-item p-pt-1 p-pb-1 p-mb-1"
          >
            <div
              class="p-grid"
            >
              <div
                class="p-col-4"
              >
                <strong>
                  Location
                </strong>
              </div>
              <div
                class="p-col-8"
              >
                <a
                  href="https://rickandmortyapi.com/api/location/20"
                  target="_blank"
                >
                  Earth (Replacement Dimension)
                </a>
              </div>
            </div>
          </div>
        </section>
        <div
          class="p-mt-auto"
        >
          <section
            class="button-bar p-mt-3"
          >
            <div
              class="p-grid"
            >
              <div
                class="p-col-4 p-text-center"
              >
                <button
                  class="p-button p-component p-button-rounded p-button-outlined p-button-icon-only"
                >
                  <span
                    class="pi pi-globe p-c"
                  />
                  <span
                    class="p-button-label p-c"
                  >
                     
                  </span>
                </button>
              </div>
              <div
                class="p-col-4 p-text-center"
              >
                <button
                  class="p-button p-component p-button-rounded p-button-outlined p-button-icon-only"
                >
                  <span
                    class="pi pi-home p-c"
                  />
                  <span
                    class="p-button-label p-c"
                  >
                     
                  </span>
                </button>
              </div>
              <div
                class="p-col-4 p-text-center"
              >
                <button
                  class="p-button p-component p-button-rounded p-button-outlined p-button-icon-only"
                >
                  <span
                    class="pi pi-external-link p-c"
                  />
                  <span
                    class="p-button-label p-c"
                  >
                     
                  </span>
                </button>
              </div>
            </div>
          </section>
          <footer
            class="p-mt-3"
          >
            <div
              class="p-grid"
            >
              <div
                class="p-col-12"
              >
                <button
                  class="p-button p-component p-button-raised p-button-lg p-text-uppercase p-text-bold"
                  style="width: 100%;"
                >
                  <span
                    class="p-button-label p-c"
                  >
                    View Episodes
                  </span>
                </button>
              </div>
            </div>
          </footer>
        </div>
        <div
          class="p-dialog-mask p-component-overlay p-dialog-center"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
